distilbert-finetune/finetune-code/finetune.py

```diff
@@ -0,0 +1,141 @@
+import argparse
+from pathlib import Path
+import os
+
+from azureml.core.run import Run
+from datasets import load_dataset
+from transformers import AutoModelForQuestionAnswering, AutoTokenizer, TrainingArguments, Trainer, DefaultDataCollator
+
+def get_args(raw_args=None):
+    parser = argparse.ArgumentParser(description="DistilBERT Fine-Tuning")
+
+    parser.add_argument("--ort", action="store_true", help="Use ORTModule")
+    parser.add_argument("--deepspeed", action="store_true", help="Use deepspeed")
+
+    args = parser.parse_args(raw_args)
+    print(f"input parameters {vars(args)}")
+    return args
+
+def preprocess_function(examples, tokenizer=None):
+    questions = [q.strip() for q in examples["question"]]
+    inputs = tokenizer(
+        questions,
+        examples["context"],
+        max_length=384,
+        truncation="only_second",
+        return_offsets_mapping=True,
+        padding="max_length",
+    )
+
+    offset_mapping = inputs.pop("offset_mapping")
+    answers = examples["answers"]
+    start_positions = []
+    end_positions = []
+
+    for i, offset in enumerate(offset_mapping):
+        answer = answers[i]
+        start_char = answer["answer_start"][0]
+        end_char = answer["answer_start"][0] + len(answer["text"][0])
+        sequence_ids = inputs.sequence_ids(i)
+
+        # Find the start and end of the context
+        idx = 0
+        while sequence_ids[idx] != 1:
+            idx += 1
+        context_start = idx
+        while sequence_ids[idx] == 1:
+            idx += 1
+        context_end = idx - 1
+
+        # If the answer is not fully inside the context, label it (0, 0)
+        if offset[context_start][0] > end_char or offset[context_end][1] < start_char:
+            start_positions.append(0)
+            end_positions.append(0)
+        else:
+            # Otherwise it's the start and end token positions
+            idx = context_start
+            while idx <= context_end and offset[idx][0] <= start_char:
+                idx += 1
+            start_positions.append(idx - 1)
+
+            idx = context_end
+            while idx >= context_start and offset[idx][1] >= end_char:
+                idx -= 1
+            end_positions.append(idx + 1)
+
+    inputs["start_positions"] = start_positions
+    inputs["end_positions"] = end_positions
+    return inputs
+
+# NOTE: Training pipeline adapted from https://huggingface.co/docs/transformers/tasks/question_answering
+def main(raw_args=None):
+    args = get_args(raw_args)
+
+    # load the SQuAD dataset from the Huggingface Datasets library
+    squad = load_dataset("squad")
+
+    # load pretrained model and tokenizer
+    tokenizer = AutoTokenizer.from_pretrained("distilbert-base-uncased")
+    model = AutoModelForQuestionAnswering.from_pretrained("distilbert-base-uncased")
+
+    # documentation: https://github.com/pytorch/ort#add-onnx-runtime-for-pytorch-to-your-pytorch-training-script
```

Comment:
Does this link work?

Reply:
Removed it, onnxruntime-training-examples itself is the best place for documentation on how to add ORT to any application :P